### DISCUSSION TITLE
Correct method of stopping main.py via firmware

### DISCUSSION THREAD

**Discussion by BrunoESP32** [micropython/micropython#13408] [ESP32] [answered]
Hi,

I'm developing an application in conjunction with the micropython firmware and I need to restart the python codes that are running.

I would like to know what would be the best solution to stop main.py or restart mp_task in the most correct way from another task.

```
  // run boot-up scripts
  pyexec_frozen_module("_boot.py", false);
  pyexec_file_if_exists("boot.py");
  if (pyexec_mode_kind == PYEXEC_MODE_FRIENDLY_REPL) {
      int ret = pyexec_file_if_exists("main.py");
      if (ret & PYEXEC_FORCED_EXIT) {
          goto soft_reset_exit;
      }
  }
```

if I achieve a soft_reset_exit or soft_reset that would be fine too.

**Comment by robert-hh** [comment]
What stops you from calling machine.soft_reset()?

**Reply by BrunoESP32** [reply]
Hi, [robert-hh](https://github.com/robert-hh)

Thanks for the answer.

The Python programs that are currently running are user programs, made on a website by them. I can even force them to call machine.soft_reset() in pyhton (I do something similar, I have a routine in python that expects a flag, from a task using a cmodule, to apply the machine.soft_reset() command), but it's a solution that I never liked.

**Comment by dlech** [comment]
Perhaps `mp_sched_vm_abort()` is what you are looking for?

**Reply by BrunoESP32** [reply]
Hi [dlech](https://github.com/dlech)

Thanks for the replay.

You are probably correct. Just to inform you, I'm using version 1.20.0 of the firmware and I saw that this function really exists, I had to enable it in ```py/mpconfig.h``` to be able to compile::

```
// Whether to support mp_sched_vm_abort to asynchronously abort to the top level.
#ifndef MICROPY_ENABLE_VM_ABORT
#define MICROPY_ENABLE_VM_ABORT (1)
#endif
```

I managed to compile the firmware, but when I call the mp_sched_vm_abort() function in main.c it is not stopping the python codes. What am I doing wrong?

Thank you very much for the suggestion.

**Reply by dlech** [reply, marked as answer]
You have to also modify the `main.c` (or other appropriate file) for whichever port you are using. The commit message (linked below) has some more info on how you would do this.

https://github.com/micropython/micropython/commit/d54208a2ff2ff8c2104597f715a586541ac6e663

**Reply by BrunoESP32** [reply]
Thanks [dlech](https://github.com/dlech)!
